File: go/windows_automation.go
// windows_automation.go
//
// Ports CircleAI.WindowsAutomation (the portable contract + logic surface):
//   Contracts.cs                 -> UiElement, UiAutomationDriver
//   InMemoryWindowsAutomation.cs -> UiAutomationEvent, InMemoryUiAutomationDriver
//   NullImplementations.cs       -> NullUiAutomationDriver
//   WindowsAutomationHelpers.cs  -> UiElement.ContainsPoint, HitTest, DumpUiElements
//
// WindowsAutomation is a UI-automation (UIA) abstraction. The real backend is a
// Win32-UIA implementation that hosts snap in for production — that native layer
// is platform-specific and NOT ported; only the driver contract, the in-memory
// virtual driver (used by tests to drive a UI without touching a desktop), the
// null fail-safe driver, and the pure hit-test/containment/dump helpers are
// ported. Native drivers implement UiAutomationDriver behind this seam, exactly
// as the C# Win32 driver implements IUiAutomationDriver.

package circleai

import (
	"context"
	"errors"
	"strconv"
	"strings"
	"sync"
)

// ---------------------------------------------------------------------------
// UiElement
// ---------------------------------------------------------------------------

// UiElement is an immutable snapshot of a single UI element: its stable id,
// accessible name, control kind, and screen rectangle. Ports the
// UiElement record.
type UiElement struct {
	ElementID string `json:"elementId"`
	Name      string `json:"name"`
	Kind      string `json:"kind"`
	X         int    `json:"x"`
	Y         int    `json:"y"`
	Width     int    `json:"width"`
	Height    int    `json:"height"`
}

// ContainsPoint reports whether (x,y) falls inside this element's rectangle.
// The right/bottom edges are exclusive, matching the C# helper exactly. Ports
// UiElementHelpers.ContainsPoint.
func (e UiElement) ContainsPoint(x, y int) bool {
	return x >= e.X && y >= e.Y && x < e.X+e.Width && y < e.Y+e.Height
}

// HitTest returns every element whose rectangle contains (x,y), preserving input
// order. Ports UiElementHelpers.HitTest.
func HitTest(elements []UiElement, x, y int) []UiElement {
	hits := make([]UiElement, 0)
	for _, e := range elements {
		if e.ContainsPoint(x, y) {
			hits = append(hits, e)
		}
	}
	return hits
}

// DumpUiElements renders elements as one debug line each, byte-for-byte matching
// the C# StringBuilder format: `id "name" kind @ (x,y) WxH\n`. Ports
// UiElementHelpers.Dump.
func DumpUiElements(elements []UiElement) string {
	var sb strings.Builder
	for _, e := range elements {
		sb.WriteString(e.ElementID)
		sb.WriteString(" \"")
		sb.WriteString(e.Name)
		sb.WriteString("\" ")
		sb.WriteString(e.Kind)
		sb.WriteString(" @ (")
		sb.WriteString(strconv.Itoa(e.X))
		sb.WriteString(",")
		sb.WriteString(strconv.Itoa(e.Y))
		sb.WriteString(") ")
		sb.WriteString(strconv.Itoa(e.Width))
		sb.WriteByte('x')
		sb.WriteString(strconv.Itoa(e.Height))
		sb.WriteByte('\n')
	}
	return sb.String()
}

// ---------------------------------------------------------------------------
// UiAutomationDriver
// ---------------------------------------------------------------------------

// UiAutomationDriver drives a UI-automation backend: enumerate elements, click,
// type, and press keys. A native Win32-UIA driver implements this in production;
// InMemoryUiAutomationDriver and NullUiAutomationDriver implement it here. Ports
// IUiAutomationDriver (the C# ValueTask methods become context.Context + error).
type UiAutomationDriver interface {
	// BackendID identifies the concrete backend ("in-memory", "null", ...).
	BackendID() string
	// Snapshot returns the current set of UI elements.
	Snapshot(ctx context.Context) ([]UiElement, error)
	// Click activates the element with the given id.
	Click(ctx context.Context, elementID string) error
	// Type sends literal text to the focused element.
	Type(ctx context.Context, text string) error
	// Key presses a named key (e.g. "Enter", "Tab").
	Key(ctx context.Context, keyName string) error
}

// ---------------------------------------------------------------------------
// UiAutomationEvent + InMemoryUiAutomationDriver
// ---------------------------------------------------------------------------

// UiAutomationEvent is raised by the in-memory driver on each interaction so a
// host can observe a virtual session. Kind is one of "click", "type", "key".
// For "click" ElementID is set; for "type"/"key" Payload carries the text/key.
// Ports the UiAutomationEvent record (ElementID / Payload optionality is modelled
// by empty strings, matching the nullable C# fields).
type UiAutomationEvent struct {
	Kind      string `json:"kind"`
	ElementID string `json:"elementId,omitempty"`
	Payload   string `json:"payload,omitempty"`
}

// InMemoryUiAutomationDriver is a real-but-virtual driver: it holds a registry of
// elements and raises observable events on Click/Type/Key without touching a real
// desktop. Ports InMemoryUiAutomationDriver. Safe for concurrent use.
type InMemoryUiAutomationDriver struct {
	mu        sync.Mutex
	elements  map[string]UiElement
	observers []func(UiAutomationEvent)
}

// NewInMemoryUiAutomationDriver constructs an empty in-memory driver.
func NewInMemoryUiAutomationDriver() *InMemoryUiAutomationDriver {
	return &InMemoryUiAutomationDriver{elements: make(map[string]UiElement)}
}

// BackendID returns "in-memory".
func (d *InMemoryUiAutomationDriver) BackendID() string { return "in-memory" }

// Register adds or replaces an element in the virtual UI, keyed by ElementID.
// Ports InMemoryUiAutomationDriver.Register.
func (d *InMemoryUiAutomationDriver) Register(el UiElement) {
	d.mu.Lock()
	defer d.mu.Unlock()
	d.elements[el.ElementID] = el
}

// Observe registers a callback invoked for every subsequent event. Ports
// InMemoryUiAutomationDriver.Observe. A nil callback is ignored.
func (d *InMemoryUiAutomationDriver) Observe(obs func(UiAutomationEvent)) {
	if obs == nil {
		return
	}
	d.mu.Lock()
	defer d.mu.Unlock()
	d.observers = append(d.observers, obs)
}

// Snapshot returns a copy of the currently registered elements.
func (d *InMemoryUiAutomationDriver) Snapshot(ctx context.Context) ([]UiElement, error) {
	if err := ctx.Err(); err != nil {
		return nil, err
	}
	d.mu.Lock()
	defer d.mu.Unlock()
	out := make([]UiElement, 0, len(d.elements))
	for _, el := range d.elements {
		out = append(out, el)
	}
	return out, nil
}

// Click raises a "click" event for elementID, rejecting blank or unknown ids —
// matching the C# ArgumentException / InvalidOperationException guards.
func (d *InMemoryUiAutomationDriver) Click(ctx context.Context, elementID string) error {
	if err := ctx.Err(); err != nil {
		return err
	}
	if strings.TrimSpace(elementID) == "" {
		return errors.New("elementId required")
	}
	d.mu.Lock()
	_, ok := d.elements[elementID]
	d.mu.Unlock()
	if !ok {
		return errors.New("unknown element '" + elementID + "'.")
	}
	d.notify(UiAutomationEvent{Kind: "click", ElementID: elementID})
	return nil
}

// Type raises a "type" event carrying text. An empty string is valid (matching
// the C# guard, which only rejects nil).
func (d *InMemoryUiAutomationDriver) Type(ctx context.Context, text string) error {
	if err := ctx.Err(); err != nil {
		return err
	}
	d.notify(UiAutomationEvent{Kind: "type", Payload: text})
	return nil
}

// Key raises a "key" event for keyName, rejecting blank names.
func (d *InMemoryUiAutomationDriver) Key(ctx context.Context, keyName string) error {
	if err := ctx.Err(); err != nil {
		return err
	}
	if strings.TrimSpace(keyName) == "" {
		return errors.New("keyName required")
	}
	d.notify(UiAutomationEvent{Kind: "key", Payload: keyName})
	return nil
}

// notify dispatches ev to every observer under a snapshot of the observer slice,
// swallowing observer panics so one bad observer cannot break dispatch — mirroring
// the C# try/catch-per-observer.
func (d *InMemoryUiAutomationDriver) notify(ev UiAutomationEvent) {
	d.mu.Lock()
	snap := make([]func(UiAutomationEvent), len(d.observers))
	copy(snap, d.observers)
	d.mu.Unlock()
	for _, o := range snap {
		func() {
			defer func() { _ = recover() }()
			o(ev)
		}()
	}
}

// ---------------------------------------------------------------------------
// NullUiAutomationDriver
// ---------------------------------------------------------------------------

// NullUiAutomationDriver is the fail-safe no-op driver: it enumerates nothing and
// accepts every command silently. Ports NullUiAutomationDriver. Use
// NullUiAutomationDriverInstance for the shared singleton.
type NullUiAutomationDriver struct{}

// NullUiAutomationDriverInstance is the shared no-op driver. Ports
// NullUiAutomationDriver.Instance.
var NullUiAutomationDriverInstance = &NullUiAutomationDriver{}

// BackendID returns "null".
func (NullUiAutomationDriver) BackendID() string { return "null" }

// Snapshot returns no elements.
func (NullUiAutomationDriver) Snapshot(context.Context) ([]UiElement, error) {
	return []UiElement{}, nil
}

// Click does nothing.
func (NullUiAutomationDriver) Click(context.Context, string) error { return nil }

// Type does nothing.
func (NullUiAutomationDriver) Type(context.Context, string) error { return nil }

// Key does nothing.
func (NullUiAutomationDriver) Key(context.Context, string) error { return nil }

// ---------------------------------------------------------------------------
// Interface guards
// ---------------------------------------------------------------------------

var _ UiAutomationDriver = (*InMemoryUiAutomationDriver)(nil)
var _ UiAutomationDriver = (*NullUiAutomationDriver)(nil)

File: c/src/windows_automation.c
/*
 * windows_automation.c — CircleAI.WindowsAutomation (C11 port).
 *
 * UiElement + UiAutomationEvent records; InMemory + Null UiAutomationDriver;
 * UiElementHelpers (ContainsPoint / HitTest / Dump).
 *
 * The in-memory driver holds an element list keyed by ElementId (Ordinal,
 * replace on dup) and an observer list. Click/Type/Key raise a UiAutomationEvent
 * to a snapshot of the observer list (so a handler that unsubscribes mid-notify
 * — not exposed here, but harmless — cannot corrupt the walk); the event is
 * borrowed for the call only. Observer "exceptions" have no analogue in C, so
 * the handler is simply invoked (the C# try/catch that logs and continues).
 *
 * Pure C11 + libc. Linear arrays (no hashtable), no pthreads.
 */

#include "circle_ai/windows_automation.h"

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <ctype.h>

/* ── per-file helpers (mirrors media.c md_*) ────────────────────────────────── */

static char *wa_strdup(const char *s) {
    if (!s) return NULL;
    size_t n = strlen(s) + 1;
    char *p = (char *)malloc(n);
    if (p) memcpy(p, s, n);
    return p;
}
static char *wa_strdup_empty(const char *s) { return wa_strdup(s ? s : ""); }

/* string.IsNullOrWhiteSpace. */
static bool wa_is_ws(const char *s) {
    if (!s) return true;
    for (const char *p = s; *p; ++p)
        if (!isspace((unsigned char)*p)) return false;
    return true;
}

/* ===========================================================================
 * UiElement record
 * =========================================================================== */

void ca_ui_element_free(ca_ui_element_t *e) {
    if (!e) return;
    free(e->element_id);
    free(e->name);
    free(e->kind);
    e->element_id = e->name = e->kind = NULL;
}
void ca_ui_element_free_array(ca_ui_element_t *arr, size_t count) {
    if (!arr) return;
    for (size_t i = 0; i < count; ++i) ca_ui_element_free(&arr[i]);
    free(arr);
}

/* Deep-copy src into dst (dst assumed uninitialised). false on OOM. */
static bool ui_element_copy(ca_ui_element_t *dst, const ca_ui_element_t *src) {
    memset(dst, 0, sizeof(*dst));
    dst->element_id = wa_strdup_empty(src->element_id);
    dst->name       = wa_strdup_empty(src->name);
    dst->kind       = wa_strdup_empty(src->kind);
    dst->x      = src->x;
    dst->y      = src->y;
    dst->width  = src->width;
    dst->height = src->height;
    if (!dst->element_id || !dst->name || !dst->kind) {
        ca_ui_element_free(dst);
        return false;
    }
    return true;
}

/* ===========================================================================
 * UiAutomationEvent record
 * =========================================================================== */

void ca_ui_automation_event_free(ca_ui_automation_event_t *ev) {
    if (!ev) return;
    free(ev->kind);
    free(ev->element_id);
    free(ev->payload);
    ev->kind = ev->element_id = ev->payload = NULL;
}

/* ===========================================================================
 * IUiAutomationDriver — InMemory + Null
 * =========================================================================== */

typedef struct {
    ca_ui_event_handler_fn handler;
    void                  *ctx;
} observer_t;

struct ca_ui_automation_driver {
    bool             is_null;
    ca_ui_element_t *elements;
    size_t           el_count, el_cap;
    observer_t      *observers;
    size_t           obs_count, obs_cap;
};

ca_ui_automation_driver_t *ca_ui_automation_driver_inmemory_create(void) {
    return (ca_ui_automation_driver_t *)calloc(1, sizeof(ca_ui_automation_driver_t));
}
ca_ui_automation_driver_t *ca_ui_automation_driver_null_create(void) {
    ca_ui_automation_driver_t *d =
        (ca_ui_automation_driver_t *)calloc(1, sizeof(*d));
    if (d) d->is_null = true;
    return d;
}
void ca_ui_automation_driver_destroy(ca_ui_automation_driver_t *drv) {
    if (!drv) return;
    for (size_t i = 0; i < drv->el_count; ++i) ca_ui_element_free(&drv->elements[i]);
    free(drv->elements);
    free(drv->observers);
    free(drv);
}
const char *ca_ui_automation_driver_backend_id(const ca_ui_automation_driver_t *drv) {
    if (!drv) return NULL;
    return drv->is_null ? "null" : "in-memory";
}

/* Find index of an element by ElementId (Ordinal). SIZE_MAX if absent. */
static size_t ui_index_of(const ca_ui_automation_driver_t *drv, const char *id) {
    for (size_t i = 0; i < drv->el_count; ++i)
        if (strcmp(drv->elements[i].element_id, id) == 0) return i;
    return (size_t)-1;
}

int ca_ui_automation_driver_register(ca_ui_automation_driver_t *drv,
                                     const ca_ui_element_t *el) {
    if (!drv || !el || drv->is_null) return -1;
    /* ElementId keys the dictionary; a real UiElement always has one. */
    if (!el->element_id) return -1;

    size_t idx = ui_index_of(drv, el->element_id);
    ca_ui_element_t copy;
    if (!ui_element_copy(&copy, el)) return -1;
    if (idx != (size_t)-1) {
        /* Dictionary set: replace in place. */
        ca_ui_element_free(&drv->elements[idx]);
        drv->elements[idx] = copy;
        return 0;
    }
    if (drv->el_count == drv->el_cap) {
        size_t nc = drv->el_cap ? drv->el_cap * 2 : 4;
        void *n = realloc(drv->elements, nc * sizeof(*drv->elements));
        if (!n) { ca_ui_element_free(&copy); return -1; }
        drv->elements = (ca_ui_element_t *)n;
        drv->el_cap = nc;
    }
    drv->elements[drv->el_count++] = copy;
    return 0;
}

int ca_ui_automation_driver_observe(ca_ui_automation_driver_t *drv,
                                    ca_ui_event_handler_fn handler, void *ctx) {
    if (!drv || !handler || drv->is_null) return -1;
    if (drv->obs_count == drv->obs_cap) {
        size_t nc = drv->obs_cap ? drv->obs_cap * 2 : 4;
        void *n = realloc(drv->observers, nc * sizeof(*drv->observers));
        if (!n) return -1;
        drv->observers = (observer_t *)n;
        drv->obs_cap = nc;
    }
    drv->observers[drv->obs_count].handler = handler;
    drv->observers[drv->obs_count].ctx     = ctx;
    drv->obs_count++;
    return 0;
}

ca_ui_element_t *ca_ui_automation_driver_snapshot(
    const ca_ui_automation_driver_t *drv, size_t *out_count) {
    if (!out_count) return NULL;
    if (!drv) { *out_count = (size_t)-1; return NULL; }
    /* NullUiAutomationDriver.SnapshotAsync -> Array.Empty. */
    if (drv->is_null || drv->el_count == 0) { *out_count = 0; return NULL; }

    ca_ui_element_t *out =
        (ca_ui_element_t *)calloc(drv->el_count, sizeof(*out));
    if (!out) { *out_count = (size_t)-1; return NULL; }
    for (size_t i = 0; i < drv->el_count; ++i) {
        if (!ui_element_copy(&out[i], &drv->elements[i])) {
            ca_ui_element_free_array(out, i);
            *out_count = (size_t)-1;
            return NULL;
        }
    }
    *out_count = drv->el_count;
    return out;
}

/* Notify: snapshot the observer list, invoke each with a borrowed event.
 * Mirrors InMemoryUiAutomationDriver.Notify (_observers.ToArray() then a
 * try/catch-per-observer loop). Observer errors have no C analogue. */
static void ui_notify(ca_ui_automation_driver_t *drv,
                      const char *kind, const char *element_id,
                      const char *payload) {
    if (drv->obs_count == 0) return;
    observer_t *snap = (observer_t *)malloc(drv->obs_count * sizeof(*snap));
    if (!snap) return;   /* cannot notify without a snapshot; drop (no throw) */
    memcpy(snap, drv->observers, drv->obs_count * sizeof(*snap));
    size_t n = drv->obs_count;

    /* A borrowed event: casts away const only so handlers get non-owning views;
     * we never free these fields (they point at caller/literal storage). */
    ca_ui_automation_event_t ev;
    ev.kind       = (char *)kind;
    ev.element_id = (char *)element_id;
    ev.payload    = (char *)payload;
    for (size_t i = 0; i < n; ++i)
        if (snap[i].handler) snap[i].handler(snap[i].ctx, &ev);
    free(snap);
}

int ca_ui_automation_driver_click(ca_ui_automation_driver_t *drv,
                                  const char *element_id) {
    if (!drv) return -1;
    if (drv->is_null) return 0;                 /* Null -> CompletedTask */
    if (wa_is_ws(element_id)) return -1;         /* ArgumentException */
    if (ui_index_of(drv, element_id) == (size_t)-1) return -1; /* Unknown element */
    ui_notify(drv, "click", element_id, NULL);
    return 0;
}

int ca_ui_automation_driver_type(ca_ui_automation_driver_t *drv,
                                 const char *text) {
    if (!drv) return -1;
    if (drv->is_null) return 0;
    if (!text) return -1;                        /* ArgumentNullException(text) */
    ui_notify(drv, "type", NULL, text);
    return 0;
}

int ca_ui_automation_driver_key(ca_ui_automation_driver_t *drv,
                                const char *key_name) {
    if (!drv) return -1;
    if (drv->is_null) return 0;
    if (wa_is_ws(key_name)) return -1;           /* ArgumentException */
    ui_notify(drv, "key", NULL, key_name);
    return 0;
}

/* ===========================================================================
 * UiElementHelpers
 * =========================================================================== */

bool ca_ui_element_contains_point(const ca_ui_element_t *el, int x, int y) {
    if (!el) return false;
    return x >= el->x && y >= el->y &&
           x < el->x + el->width && y < el->y + el->height;
}

ca_ui_element_t *ca_ui_element_hit_test(const ca_ui_element_t *els, size_t n,
                                        int x, int y, size_t *out_count) {
    if (!out_count) return NULL;
    if (n == 0) { *out_count = 0; return NULL; }
    if (!els) { *out_count = (size_t)-1; return NULL; }

    size_t *idx = (size_t *)malloc(n * sizeof(size_t));
    if (!idx) { *out_count = (size_t)-1; return NULL; }
    size_t hits = 0;
    for (size_t i = 0; i < n; ++i)
        if (ca_ui_element_contains_point(&els[i], x, y)) idx[hits++] = i;

    if (hits == 0) { free(idx); *out_count = 0; return NULL; }
    ca_ui_element_t *out = (ca_ui_element_t *)calloc(hits, sizeof(*out));
    if (!out) { free(idx); *out_count = (size_t)-1; return NULL; }
    for (size_t i = 0; i < hits; ++i) {
        if (!ui_element_copy(&out[i], &els[idx[i]])) {
            ca_ui_element_free_array(out, i);
            free(idx);
            *out_count = (size_t)-1;
            return NULL;
        }
    }
    free(idx);
    *out_count = hits;
    return out;
}

/* Grow-or-fail append into a heap buffer (buf/len/cap by pointer). false on OOM. */
static bool dump_append(char **buf, size_t *len, size_t *cap, const char *s) {
    size_t sl = strlen(s);
    if (*len + sl + 1 > *cap) {
        size_t nc = *cap ? *cap : 64;
        while (*len + sl + 1 > nc) nc *= 2;
        void *n = realloc(*buf, nc);
        if (!n) return false;
        *buf = (char *)n;
        *cap = nc;
    }
    memcpy(*buf + *len, s, sl + 1);
    *len += sl;
    return true;
}

char *ca_ui_element_dump(const ca_ui_element_t *els, size_t n) {
    if (n > 0 && !els) return NULL;
    size_t cap = 128, len = 0;
    char *buf = (char *)malloc(cap);
    if (!buf) return NULL;
    buf[0] = '\0';
    for (size_t i = 0; i < n; ++i) {
        const ca_ui_element_t *e = &els[i];
        /* <ElementId> "<Name>" <Kind> @ (<X>,<Y>) <Width>x<Height>\n
         * The int coordinates go through a fixed scratch; the strings append
         * directly so we never truncate a long ElementId/Name/Kind. */
        char coords[128];
        if (!dump_append(&buf, &len, &cap, e->element_id ? e->element_id : "") ||
            !dump_append(&buf, &len, &cap, " \"") ||
            !dump_append(&buf, &len, &cap, e->name ? e->name : "") ||
            !dump_append(&buf, &len, &cap, "\" ") ||
            !dump_append(&buf, &len, &cap, e->kind ? e->kind : "")) {
            free(buf);
            return NULL;
        }
        snprintf(coords, sizeof(coords), " @ (%d,%d) %dx%d\n",
                 e->x, e->y, e->width, e->height);
        if (!dump_append(&buf, &len, &cap, coords)) { free(buf); return NULL; }
    }
    return buf;
}
